src/replays/database_replay_provider/database_replay_provider.ts

```diff
@@ -0,0 +1,268 @@
+import { FileRepository } from "@database/repositories/file_repository";
+import { GameRepository } from "@database/repositories/game_repository";
+import { PlayerRepository } from "@database/repositories/player_repository";
+import type { Database, FileRecord, NewFile, NewGame, NewPlayer, PlayerRecord } from "@database/schema";
+import type { FileLoadResult, FileResult, Progress, ReplayProvider } from "@replays/types";
+import type { StadiumStatsType, StatsType } from "@slippi/slippi-js";
+import { SlippiGame } from "@slippi/slippi-js";
+import log from "electron-log";
+import * as fs from "fs-extra";
+import type { Kysely } from "kysely";
+import path from "path";
+
+import { extractPlayerNames } from "../file_system_replay_provider/extract_player_names";
+import { inferStartTime } from "./infer_start_time";
+import { mapGameRecordToFileResult } from "./record_mapper";
+
+const INSERT_REPLAY_BATCH_SIZE = 200;
+
+export class DatabaseReplayProvider implements ReplayProvider {
+  constructor(private readonly db: Kysely<Database>) {}
+
+  public async loadFile(filePath: string): Promise<FileResult> {
+    const filename = path.basename(filePath);
+    const folder = path.dirname(filePath);
+
+    let gameRecord = await GameRepository.findGameByFolderAndFilename(this.db, folder, filename);
+
+    if (!gameRecord) {
+      // Add the game if it doesn't already exist in our database
+      const { fileId } = await this.insertNewReplayFile(folder, filename);
+
+      // TODO: Figure out how to return the game record directly after adding
+      // to avoid needing to do another database query
+      gameRecord = await GameRepository.findGameByFileId(this.db, fileId);
+      if (!gameRecord) {
+        throw new Error(`Could not find file ${fileId} at path: ${filePath}`);
+      }
+    }
+
+    const playerRecords = await PlayerRepository.findAllPlayersByGame(this.db, gameRecord._id);
+    return mapGameRecordToFileResult(gameRecord, playerRecords);
+  }
+
+  public async loadFolder(folder: string, onProgress?: (progress: Progress) => void): Promise<FileLoadResult> {
+    // If the folder does not exist, return empty
+    if (!(await fs.pathExists(folder))) {
+      return {
+        files: [],
+        fileErrorCount: 0,
+        totalBytes: 0,
+      };
+    }
+
+    // Add new files to the database and remove deleted files
+    await this.syncReplayDatabase(folder, onProgress, INSERT_REPLAY_BATCH_SIZE);
+
+    const [gameRecords, totalBytes] = await Promise.all([
+      GameRepository.findGamesByFolder(this.db, folder),
+      FileRepository.findTotalSizeByFolder(this.db, folder),
+    ]);
+
+    const players = await Promise.all(
+      gameRecords.map(async ({ _id: gameId }): Promise<[number, PlayerRecord[]]> => {
+        const playerRecords = await PlayerRepository.findAllPlayersByGame(this.db, gameId);
+        return [gameId, playerRecords];
+      }),
+    );
+    const playerMap = new Map(players);
```

Comment:
This seems a bit slow on a relatively large db.

Reply:
Fixed in 43141c38. :)